src/loader/start_info.rs

```diff
@@ -0,0 +1,298 @@
+ *  4 +----------------+
+ *    | version        | Version of this structure. Current version is 1. New
+ *    |                | versions are guaranteed to be backwards-compatible.
+ *  8 +----------------+
+ *    | flags          | SIF_xxx flags.
+ * 12 +----------------+
+ *    | nr_modules     | Number of modules passed to the kernel.
+ * 16 +----------------+
+ *    | modlist_paddr  | Physical address of an array of modules
+ *    |                | (layout of the structure below).
+ * 24 +----------------+
+ *    | cmdline_paddr  | Physical address of the command line,
+ *    |                | a zero-terminated ASCII string.
+ * 32 +----------------+
+ *    | rsdp_paddr     | Physical address of the RSDP ACPI data structure.
+ * 40 +----------------+
+ *    | memmap_paddr   | Physical address of the (optional) memory map. Only
+ *    |                | present in version 1 and newer of the structure.
+ * 48 +----------------+
+ *    | memmap_entries | Number of entries in the memory map table. Zero
+ *    |                | if there is no memory map being provided. Only
+ *    |                | present in version 1 and newer of the structure.
+ * 52 +----------------+
+ *    | reserved       | Version 1 and newer only.
+ * 56 +----------------+
+ *
+ * The layout of each entry in the module structure is the following:
+ *
+ *  0 +----------------+
+ *    | paddr          | Physical address of the module.
+ *  8 +----------------+
+ *    | size           | Size of the module in bytes.
+ * 16 +----------------+
+ *    | cmdline_paddr  | Physical address of the command line,
+ *    |                | a zero-terminated ASCII string.
+ * 24 +----------------+
+ *    | reserved       |
+ * 32 +----------------+
+ *
+ * The layout of each entry in the memory map table is as follows:
+ *
+ *  0 +----------------+
+ *    | addr           | Base address
+ *  8 +----------------+
+ *    | size           | Size of mapping in bytes
+ * 16 +----------------+
+ *    | type           | Type of mapping as defined between the hypervisor
+ *    |                | and guest.
+ * 20 +----------------|
+ *    | reserved       |
+ * 24 +----------------+
+ *
+ * The address and sizes are always a 64bit little endian unsigned integer.
+ */
+
+// Rust definitions needed to enable PVH boot protocol
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Default)]
+pub struct hvm_start_info {
+    pub magic: u32,
+    pub version: u32,
+    pub flags: u32,
+    pub nr_modules: u32,
+    pub modlist_paddr: u64,
+    pub cmdline_paddr: u64,
+    pub rsdp_paddr: u64,
+    pub memmap_paddr: u64,
+    pub memmap_entries: u32,
+    pub reserved: u32,
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, Default)]
+pub struct hvm_memmap_table_entry {
+    pub addr: u64,
+    pub size: u64,
+    pub type_: u32,
+    pub reserved: u32,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn bindgen_test_layout_hvm_start_info() {
```

Comment:
Is this file generated with bindgen? If so, please add a comment like in the other generated files (elf.rs, bootparam.rs). If partially, please add instructions on how to replicate it.

Reply:
Partially. I used bindgen initially but the output is not very legible, so I opted for manually copying the comments and the definitions I needed into start_info.rs. The tests I added later when trying to figure out the coverage failure.

I'll add instructions.

Reply:
Generated `start_info.rs` again using `bindgen`, grabbing the `start_info.h` header from Xen tagged release 4.13.0. Made these changes to the generated start_info.rs:

- Kept the original Citrix copyright stanza.
- Listed bindgen command and upstream commit used to generate the file as requested by @aghecenco.
- Added comment detailing purpose of the definitions as requested by @andreeaflorescu.
- Kept comments from C header with structure layout and explanations.
- Included the hvm_modlist_entry definition which is needed for supporting initrd in PVH.
- Added Default to the list of traits that are automatically derived for each of the definitions.
- Grouped definitions in the default module and moved all the tests to test module.
- Edited the file for clarity, trimming unnecessary definitions.

Please see newest version and let me know if anything is missing.